march_imu_manager/src/WirelessMaster.cpp

```diff
@@ -0,0 +1,249 @@
+
+    if (!this->m_control->openPort(wirelessMasterPort->portName().toStdString(), wirelessMasterPort->baudrate()))
+    {
+        ROS_FATAL_STREAM("Failed to open port " << *wirelessMasterPort);
+        return -1;
+    }
+
+    ROS_DEBUG("Getting XsDevice instance for wireless master...");
+    this->m_master = this->m_control->device(wirelessMasterPort->deviceId());
+
+    ROS_DEBUG("Attaching callback handler for master...");
+    this->m_master->addCallbackHandler(this);
+
+    return 0;
+}
+
+int WirelessMaster::configure(const int updateRate, const int channel)
+{
+    if (this->m_master && !this->m_master->gotoConfig())
+    {
+        ROS_FATAL("Failed to go to config mode");
+        return -1;
+    }
+
+    ROS_DEBUG("Getting the list of the supported update rates...");
+    const XsIntArray supportedUpdateRates = this->m_master->supportedUpdateRates();
+
+    std::ostringstream updateRates;
+    for (const int updateRate : supportedUpdateRates)
+    {
+        updateRates << updateRate << " ";
+    }
+    ROS_DEBUG_STREAM("Supported update rates: " << updateRates.str());
+
+    const int newUpdateRate = findClosestUpdateRate(supportedUpdateRates, updateRate);
+
+    ROS_DEBUG_STREAM("Setting update rate to " << newUpdateRate << " Hz...");
+    if (!this->m_master->setUpdateRate(newUpdateRate))
+    {
+        ROS_FATAL_STREAM("Failed to set update rate");
+        return -1;
+    }
+
+    ROS_DEBUG("Disabling radio channel if previously enabled...");
+    if (this->m_master->isRadioEnabled())
+    {
+        if (!this->m_master->disableRadio())
+        {
+            ROS_FATAL_STREAM("Failed to disable radio channel");
+            return -1;
+        }
+    }
+
+    ROS_DEBUG_STREAM("Setting radio channel to " << channel << " and enabling radio...");
+    if (!this->m_master->enableRadio(channel))
+    {
+        ROS_FATAL_STREAM("Failed to set radio channel");
+        return -1;
+    }
+
+    return 0;
+}
+
+void WirelessMaster::waitForConnections(const size_t connections)
+{
+    std::unique_lock<std::mutex> lck(this->m_mutex);
+
+    auto hasConnections = [this, connections]
+    {
+        return this->m_connectedMtws.size() == connections;
+    };
+    this->m_cv.wait(lck, hasConnections);
+}
+
+bool WirelessMaster::startMeasurement()
+{
+    return this->m_master && this->m_master->gotoMeasurement();
+}
+
+bool WirelessMaster::isMeasuring() const
+{
+    return this->m_master && this->m_master->isMeasuring();
+}
+
+void WirelessMaster::update()
+{
+    for (const auto& mtw : this->m_connectedMtws)
+    {
+        if (mtw.second->dataAvailable())
+        {
+            const XsDataPacket* packet = mtw.second->getOldestPacket();
+
+            if (packet->containsCalibratedData())
+            {
+                sensor_msgs::Imu imu_msg;
+
+                imu_msg.header.frame_id = "imu_" + mtw.second->getId().toString().toStdString();
```

Comment:
Hier moet niet de ID van de IMU komen, maar de ID van het frame waar de sensor bij hoort. Aangezien de IMU nog niet is toegevoegd aan de URDF heeft dit nog geen zin. Voor nu zou de frame_id dus leeggelaten moeten worden.

Reply:
I'll remove it for now